src/resolve/resolved-mdns-browse-services.c

```diff
@@ -0,0 +1,697 @@
+ * Adding 1 to TTL to compensate for clamped TTL. */
+#define MDNS_MAINTENANCE_NEXT(usec,ttl,base_perc) (usec_add(usec, (base_perc) * (1 + ttl) * USEC_PER_SEC))
+
+/* RFC 6762 section 5.2 - A random variation of 2% of the record TTL should
+ * be added to maintenance queries. */
+#define MDNS_MAINTENANCE_JITTER ((double) rand() / RAND_MAX * 0.02)
+
+#define MDNS_80_PERCENT 0.8
+#define MDNS_5_PERCENT 0.05
+
+typedef struct mDnsGoodbyeParams mDnsGoodbyeParams;
+
+struct mDnsGoodbyeParams {
+        mDnsServiceSubscriber *ss;
+        int owner_family;
+};
+
+mDnsGoodbyeParams *mdns_goodbye_params_free(mDnsGoodbyeParams *p);
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(mDnsGoodbyeParams*, mdns_goodbye_params_free);
+
+static void mdns_find_service_from_query(mDnsServices **service, mDnsServiceSubscriber *ss, DnsQuery *q) {
+        assert(ss);
+
+        /* Find the service that owns the query. */
+        LIST_FOREACH(mdns_services, s, ss->mdns_services) {
+                if (s->query == q) {
+                        *service = s;
+                        return;
+                }
+        }
+        *service = NULL;
+}
+
+static void mdns_maintenance_query_complete(DnsQuery *q) {
+        _cleanup_(mdns_service_subscriber_unrefp) mDnsServiceSubscriber *ss = NULL;
+        _cleanup_(dns_query_freep) DnsQuery *query = q;
+        mDnsServices *service = NULL;
+        int  r;
+
+        assert(query);
+        assert(query->manager);
+
+        ss = mdns_service_subscriber_ref(query->manager->dns_service_subscriber);
+        if (!ss)
+                return;
+
+        if (query->state != DNS_TRANSACTION_SUCCESS) {
+                r = 0;
+                goto finish;
+        }
+
+        r = dns_answer_match_key(query->answer, ss->key, NULL);
+        if (r <= 0)
+                goto finish;
+
+        r = mdns_subscriber_lookup_cache(ss, query->answer_family);
+
+finish:
+        if (r < 0)
+                log_error_errno(r, "mDNS maintenance query complete failed: %m");
+
+        mdns_find_service_from_query(&service, ss, query);
+        if (service)
+                service->query = NULL;
+}
+
+static int mdns_maintenance_query(sd_event_source *s, uint64_t usec, void *userdata) {
+        mDnsServices *service = NULL;
+        _cleanup_(dns_query_freep) DnsQuery *q = NULL;
+        int r;
+
+        assert(userdata);
+
+        service = userdata;
+
+        switch (service->rr_ttl_state) {
+        case MDNS_TTL_80_PERCENT:
+                service->rr_ttl_state = MDNS_TTL_85_PERCENT;
+                break;
+
+        case MDNS_TTL_85_PERCENT:
+                service->rr_ttl_state = MDNS_TTL_90_PERCENT;
+                break;
+
+        case MDNS_TTL_90_PERCENT:
+                service->rr_ttl_state = MDNS_TTL_95_PERCENT;
+                break;
+
+        case MDNS_TTL_95_PERCENT:
+                service->rr_ttl_state = MDNS_TTL_100_PERCENT;
+                break;
```

Comment:
if this progresses linearly it probably shouldn't be treated like a state engine, but just as a progress counter. i.e. so that you can just increase the counter to get to the next step

Reply:
changed to incrementing counter